### DISCUSSION TITLE
Get the `start` property of a sslice.

### DISCUSSION THREAD

**Discussion by woojamon** [koka-lang/koka#459] [General]
I see in sslice.kk that `sslice` has a `start` property, but there doesn't seem to be a way in user code to access it directly or by pattern matching against `Sslice`. Am I missing something?

**Comment by TimWhiting** [comment]
Because it is representation specific (depending on the underlying platform it could be utf8 or utf16, etc) and the start property is a byte offset it is intentionally hidden. What do you need it for? You should be able to do what you want with truncate, extend, and advance as well as some other slice functions defined in core/slice. If you want to work at the uint8 level, you'll have to wait for the bytes module.

**Reply by TimWhiting** [reply]
In other words, an abstract datatype is not constructable or destructable (accessing properties or matching) outside the module in which it is defined. It is intended to be only used as the module's interface allows. In this case because the underlying datatype is backend / representation specific, it is not safe to modify the start / end properties, because you could accidentally move them to somewhere within a utf16 character and unintentionally create a bad representation).

**Reply by TimWhiting** [reply]
Additionally Koka's character type is not a uint8 type either, it represents a full utf character.

**Reply by woojamon** [reply]
Thanks, I need it for inspection/debugging.

I was writing some tests with assertions that were comparing some `sslice`s and couldn't understand why two slices with otherwise identical strings weren't comparing equally until I remembered that `sslices`s keep a pointer to a start position in the string they were created on. 

For example if I created a `sslice` from `"abcd".slice.next.unjust.snd` and compared it with `"bcd".slice` they would not be equal. It wasn't sufficient to test string equality either, because the code I was testing works against `sslice`s.

So what I wanted to do was create a `show` for `sslice` and then print it with `trace` so that I could more easily spot this issue in future tests and fix the test setup.
```koka
pub fun sslice/show(ss)
  match ss
    Sslice(str, start, len) -> "Sslice(str=\"" ++ str ++ "\", start=" ++ start ++ ", len=" ++ len ++ ")"
```
or
```koka
pub fun sslice/show(ss)
  "Sslice(str=\"" ++ ss.string ++ "\", start=" ++ ss.start ++ ", len=" ++ ss.count ++ ")"
```

**Reply by TimWhiting** [reply]
Hmm, yes this is an interesting issue. I think we should change equality on slices to first check if the strings are identical (using pointer equality) and then if not, convert it to strings prior to doing equality check. I'll create a PR and talk to Daan about it.

**Reply by woojamon** [reply]
Thanks!

Are you thinking `start` still be part of the equality check?

**Reply by TimWhiting** [reply]
No, unless the strings are identical pointers.

This would be the new definition
```koka
pub inline fun (==)(l1: sslice, l2: sslice ): bool
  match (l1, l2)
    (Sslice(s1, i1, f1), Sslice(s2, i2, f2)) ->
      if identical(s1, s2) then i1 == i2 && f1 == f2
      else l1.string == l2.string

pub inline fun (!=)(l1: sslice, l2: sslice ): bool
  match (l1, l2)
    (Sslice(s1, i1, f1), Sslice(s2, i2, f2)) ->
      if identical(s1, s2) then i1 != i2 || f1 != f2
      else l1.string != l2.string
```
This way, slices are equal if their contents are equal.

**Reply by woojamon** [reply]
What would be the result of the equalities below under the new definition?

```koka
val slice1 = "abcd".slice.advance(1)
val slice2 = "abcd".slice.advance(2)
slice1 == slice2
```
```koka
val str = "abcd"
val slice1 = str.slice.advance(1)
val slice2 = str.slice.advance(2)
slice1 == slice2
```

**Reply by TimWhiting** [reply]
Both false

**Reply by woojamon** [reply]
Hm ok I must not understand string pointer equality then, or at least how koka compiles it. It seems to me like dropping the `start` equality if the string pointers aren't identical could affect user code behavior in some cases. How about the following?

```koka
val str1 = "abcd"
val slice1 = str1.slice.advance(1)

val str2 = "abcd"
val slice2 = str2.slice.advance(2)

slice1 == slice2
```

```koka
val str1 = "aabcd".tail
val slice1 = str1.advance(1)

val str2 = "abcd"
val slice2 = str2.slice.advance(2)

slice1 == slice2
```

**Reply by TimWhiting** [reply]
That is because equality for slices should be based on the *contents* of the slice not the pointers and offsets. Essentially it is an optimized `slice1.string == slice2.string`. It shouldn't depend on what strings you start with and what operations you did to get there. Your examples should both return false. The start and end are internal representation, and don't determine if the values are equal. If you want to check for identical slices that would be different, but still all your examples would return false.

**Reply by TimWhiting** [reply]
Okay, I see at least one situation where my definition fails still to live up to my definition of equality based on contents.

**Reply by woojamon** [reply]
Well, and my example-making skills probably suck too, but at the end of the day I do think I would want the following to be **not equal**, regardless of internal string pointer stuff that I also don't understand 😅:
```koka
Sslice("abcd", 1, 4)
Sslice("abcd", 2, 4)
```

**Reply by woojamon** [reply]
I guess what I'm saying is if I had slices and wanted to check string equality I could just go ahead and do it `slice1.string == slice2.string`, so I would expect `slice1 == slice2` to be slice-equality and do more than just string equality internally. 

I'm also willing to concede that I'm just not conceiving of slices in the way that you and Daan as language designers are. 👍

**Reply by TimWhiting** [reply]
Yes, your example would be **not equal**. It does do more than string equality, since it has an optimized path if the underlying string is the same, but at the end of the day it should use content equality which is the same as string equality.

I think the example that you want is that you want these to be non-equal even though their contents are equal.
```
val str = "abcabc"
Sslice(str, 0, 3) // .string == "abc"
Sslice(str, 3, 6) // .string == "abc"
```
However, that does not make sense since the only view that a user has into the slice is based on the contents, and cannot retrieve the offset (at least in bytes, I guess you could keep track of the number of characters you are into the string). We could probably add a `eq-slice` function or `===` that would check equality of the fields as was done previously. I'm just wondering when it would be used other than to optimize an `==` when you know that the backing string is always referentially equal - which I'm proposing we do already as part of the `==` method.

If you want to check referential/structural equality, that would be a different function, and not specific to slices. (I added it in the PR, but it doesn't do deep structural equality for value types). Though again we could probably add a `===` to slices in particular. I just don't see the use case.

**Reply by woojamon** [reply]
Wow! Great example 👍

So I think this part: "the only view that a user has into the slice is based on the contents", is certainly true, but the functions in std/core/sslice have the deeper view where they _can_ see and work with `start` and `len`.

So my thought (which may be moot) is that a user would come across something where they are parsing a file or something and doing a bunch of slice work, and somewhere wanting to test some slices against each other and end up at runtime having a scenario like your example...

```koka
if slice1 == slice2 // "abc" == "abc" but underneath its really something like Sslice(str, 0, 3) != Sslice(str, 3, 6)
  ...
```
then later on using some functions from std/core/sslice and just getting completely unexpected results because the underlying slices really are different.

**Reply by TimWhiting** [reply]
Okay, I can maybe see that. I do think that it needs to be a different function though. Any ideas for naming the function?

**Reply by TimWhiting** [reply]
Also, I think in the parsing use case usually you'd be keeping track of the offset separately. Maybe there needs to be a version of slice that keeps track of character offsets, in addition to the bytes (which are important for fast access in the C implementation).

**Reply by woojamon** [reply]
> Any ideas for naming the function?

That is one of the hard parts of programming ha.

I know in some other functional-oriented, immutable-by-default programming languages `==` is used for structural equality and `===` for reference equality, but it maybe that koka is already committed to a certain concept of equality for `==`. 

It seems like the topic now though is between something like `sslice/slice-eq` and `sslice/string-eq`, but because a user can easily check string equality of slices via `.string`, I would advocate that `==` for slices checks something a little "`sslice`ier". (Of course, I've only been playing around with slices for a few hours, where as language designers you and Daan have a much stronger vision of what slices should be for users, so I approach the issue open-handed.)

**Reply by woojamon** [reply]
> Also, I think in the parsing use case usually you'd be keeping track of the offset separately. Maybe there needs to be a version of slice that keeps track of character offsets, in addition to the bytes (which are important for fast access in the C implementation).

Give me a while to flesh out my parsing codebase a little more and I'll share it with you so you can see how I'm using `sslice`.

**Reply by woojamon** [reply]
Sorry to cause any "confusion by discussion". 

_I'm actually quite happy with the current way slice equality works,_ just wanted to be able to inspect the value of `start` property.

**Reply by TimWhiting** [reply]
Do you want start (byte offset) or the offset in utf characters? `string.list` / `char` in Koka is utf encoded, and has the potential to differ based on the platform (C / JS / C#).

**Reply by woojamon** [reply]
I think offset in UTF characters would be more intuitive for users long term (although for my current purposes either works).

The reason goes back to what you said earlier about the view users have into slices being based on their (string) contents. The functions in the sslice module also seem to strongly reinforce the idea, since many of them "work" in terms of chars and strings from the user's perspective (`advance`, `next`, `count`, `drop`, `foreach` etc). So whether working with a hardcoded a string, or parsing file contents, or receiving console input or what-have-you, regardless of the underlying byte layout, its just natural to think in terms of operating on the string input at the character level, and the public sslice functions express the same idea to boot. (Although I guess someone could want to parse a file at the byte level, but at the moment they'd have a tough time using the sslice module to do it.) It seems to make sense then that the user would want to know "where (what character index) does this slice start on its underlying string". 
This way also keeps `start` privately hidden behind a layer of abstraction from the user, so that language maintainers can refactor the underlying byte-driven implementation in the future with much less concern about affecting the public behavior and compatibility with existing user code.

As far as the platform concerns you mentioned, what would be an example where that could bite someone?

**Reply by TimWhiting** [reply]
The byte offset would obviously be different between platforms. So a web server in c / web client compiled to javascript could have issues if they send strings and offsets to each other. Koka compiled to Javascript uses utf16, but Koka compiled to C uses utf8. I agree that we should have utf character offsets available for slices, the question is whether to calculate them on demand, `O(n)` or take up the extra storage space. In the Koka compiler's lexer for example, we keep track of the offset as part of the lexer's state to associate tokens to positions in the file, especially when you take into account line breaks and wanting to know column/row offsets as well, it makes most sense to keep track of that outside of a slice. Ultimately it is up to Daan, but would having an `O(n)` method to get the character offset work for you?

**Reply by woojamon** [reply]
Oh, thanks Tim, it makes sense the cross platform communication scenario would cause an issue if transferring byte offsets, which I think lends further weight to using char offset.

However, as for which solution works for me, I would say _**any**_ of 
1. inspecting byte offset, 
2. inspecting a char-offset property, or 
3. calling a char-offset method (`O(n)` or otherwise) 

works for me.

Because the fundamental problem I would like to be able to solve is: "When koka says that two slices are not equal, _I want to be able to discover **why** they're not equal_." Therefore, literally _**any solution**_ that allows me to do that would work for me. 
("Parsing" just so happened to be the context I was working in, I could actually care less whether increased discoverability helps me parse strings or not.)

Now, I know a developer could just do what I did and come to the repo and look at the sslice module to see how sslice equality is computed, but I just like inspectability because it allows me to make a function that "pretty prints" a sslice pattern to console when a test fails, so I can just look at the output and say "oh yeah these two slices are different right there, let me go account for that". 

Granted, any solution has potential to be misused or misunderstood by developers (maybe it could even be an `unsafe-` kind of method?), but I trust you and Daan to make the call as language designers as to what the most balanced solution would be. 

Thank you for your time in talking this out with me!
